internal/server/polling/server.go

```diff
@@ -97,9 +100,75 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
 		return fmt.Errorf("failed to list pull requests: %w", err)
 	}
 
+	// Log the PRs.
 	for _, pr := range prs {
 		s.log.Info("pull request", "pr", pr)
 	}
 
+	labels := map[string]string{
+		bbp.LabelKey: bbp.LabelValue,
+	}
+
+	// List Terraform objects, created by BBP.
+	tfList, err := s.listTerraformObjects(ctx, tf, labels)
+	if err != nil {
+		return fmt.Errorf("failed to list Terraform objects: %w", err)
+	}
+
+	// Create a list of objects to delete.
+	oldTFs := []*infrav1.Terraform{}
+
+	for _, branchTF := range tfList {
+		shouldDeleteTF := true
+
+		for _, pr := range prs {
+			if fmt.Sprintf("%d", pr.Number) == branchTF.Labels[bbp.LabelPRIDKey] {
+				shouldDeleteTF = false
+				break
+			}
+		}
+
+		if shouldDeleteTF {
+			oldTFs = append(oldTFs, branchTF)
+		}
+	}
+
+	// Delete TF objects without corresponding PRs.
+	for _, oldTF := range oldTFs {
+		err = s.deleteTerraform(ctx, oldTF)
```

Comment:
I think we can move it up into the previous double-loop. Right now we have a quadratic double-loop that allocates and constantly resize a new slice, just to loop over all items and call delete.

Even setting the capacity of `oldTFs` to `len(tfList)` can save us a lot of resizing. Or simply just get rid of this slice.